rfcs/build-system/05-folder-organization.md

```diff
@@ -0,0 +1,65 @@
+# RFC: Fluent UI Repo Folder Organization
+
+---
+
+Contributors: Fluent UI React Build team, @justSlone
+
+## Summary
+
+This propsoal covers 2 topics:
+1. organization of rfcs folder
+2. organization (or removal) of specs folder
+
+Overall this RFC proposes that we organize our repo by NPM package names under the @fluentui namespace. 
+These names today are `react`, `react-components`, `react-northstar`, and `web-components`. Organizing by NPM package name, while not necessarily ideal and may change over time, has the benefit of making it easy to follow from NPM to the appropriate source code in GitHub. This also scales well if we add additional projects to the repo (as long as they are under the @fluentui namespace on NPM). 
+
+## Background
+
+Some useful background is that we have heard from contributors that the current organization of the rfcs and specs folders at the root of the 
+repo is confusing to contributors working on web-components. The RFCs and Specs don't mention clearly what project they are for. 
+
+## Problem statement
+
+With the addition of vNext to the repo we now have 4 actual projects in our repo: 
+`@fluentui/react`, `@fluentui/react-northstar`, `@fluentui/web-components`, and more recently `@fluentui/react-components`.
+
+This leads to a fairly confusing and difficult to navigate repo. In particular we have some folders we are actively using, such as rfcs and specs which could apply to any of the 4 projects, but in fact only really apply to @fluentui/react-components. This is potentially confusing to contributors visiting our repo, and generally a ramp-up headache. 
+
+## Detailed Design or Proposal
+
+1. Within the rfcs folder we should create a new subfolder called `react-components` and move all existing rfcs to this folder. 
+   - if needed additional `react`, `react-northstar`, or `web-components` RFCs can be created
+   - RFCs that apply to the whole repo can go in a folder called `shared` 
+2. Remove the specs folder and move any specs to the relevant packages folders
+   - We should be keeping specs up to date under the relevant package folders. 
+   - Associating specs with particular packages scales better in a large repo such as this. 
+   - The reason we don't do this for RFCs is that they span packages (and as is in the case of this RFC, sometimes span projects). 
+
+### Pros and Cons
+
+Pros:
+- Better organization of repo, more clear for contributors
+- Improves the scale of the repo, could allow more projects in the future, or at least makes working with the projects we have more pleasant
+
+Cons:
+- Large file moves cause merge conflicts
+
+## Discarded Solutions
```

Comment:
I believe one solution that also comes to my mind is converting our RFC process into using Github Issues as pretty much all open source projects do.

As I understand, RFCs are supposed to be perishable, they are "requests for comments" so once comments are provided and solution has been agreed upon that solution should be implemented and there is no further need for an RFC.

If we want persistent source of truth I'd prefer to have [ADRs (Architecture Decision Record)](https://18f.gsa.gov/2021/07/06/architecture_decision_records_helpful_now_invaluable_later/) or have it written within the repo Wiki.

Currently, as we go on, we will have a growing collection of RFC documents within the repo that are never removed despite the fact they may no longer be relevant.

Opting into using Issues for them allows easier tracking, the ability to cross-reference them and link to other Issues/PRs/Projects easily or even label them efficiently to resolve the current problem of which project the RFC belongs to.

Reply:
>  believe one solution that also comes to my mind is converting our RFC process into using Github Issues as pretty much all open source projects do.

It depends case by case. some use issues some dont. You can better comment when you create document and have conversation rather than very long issue thread that may get out off hand.


> Currently, as we go on, we will have a growing collection of RFC documents within the repo that are never removed despite the fact they may no longer be relevant.

Where did you get this assumption? If RFC/Spec is no longer valid it should be removed

> If we want persistent source of truth I'd prefer to have ADRs

Right, the RFC are meant to be that source of truth as well, just not named ADRs

> within the repo Wiki.

wiki functionality is problematic on Github, hard to track and other issues. We decided to not use it for those main reasons


> Opting into using Issues for them allows easier tracking, the ability to cross-reference them and link to other Issues/PRs/Projects easily or even label them efficiently to resolve the current problem of which project the RFC belongs to.

FWIW all those things you mentioned does apply for current workflow as well - there is document with proper git history - thus 100% trackable

Reply:
> Where did you get this assumption? If RFC/Spec is no longer valid it should be removed

Will people have enough context to go over all the RFCs periodically and decide which ones are outdated and should be removed? I presume most people will just create new RFCs and seldom revisit the existing ones.

> Right, the RFC are meant to be that source of truth as well, just not named ADRs

But if they are supposed to be the source of truth then they have to be persistent so they cannot be removed.

I understand that they may serve the role of ADRs but then it can lead to confusion from people to assume request for comments serve as an architecture decision documents. But that's just my own subjective opinion.

> FWIW all those things you mentioned does apply for current workflow as well - there is document with proper git history - thus 100% trackable

Yes, the history is persistent but that's about it, Issues themselves provide a lot more functionality and benefits mentioned above in addition to historical trackability over markdown files within the repository.

Reply:
We started out doing RFCs as github issues, but the discussion process was extremely unwieldy, so starting with #16574 we switched to using files. As mentioned in that PR description, checking in RFCs is similar to what many other major libraries do, including React and npm. I think it's working pretty well overall, but definitely open to refinements.

One area for improvement is I agree that it would be good to have more clarity about the purpose and lifetime of RFCs. That should be documented in `rfcs/README.md`.

Reply:
I have to agree here in saying that I very much prefer to review RFCs as documents in a PR rather than as issues. The lifetime of RFCs is a valid point though and I think we should probably go and review all RFCs once we get closer to official release. We should probably also document which version they refer to at that point.

Reply:
Yeah, I agree on the commenting in PRs, I feel it has helped keep RFC review discussions productive by allowing sub-conversations regarding particular aspects of the RFC. 

You know, I was originally resistant to the idea of periodically reviewing RFCs (and was more inline with the ADR approach of not touching old ADRs), but you know it might be a reasonable thing to do alongside major releases. There is a lot of value in a well-curated set of reasoning for the decisions. Investing in this now would set us up for easier maintenace of the project in the long term.

In the end, I think our RFCs are a bit closer to ADRs than just "hey, I had an idea, give me some comments". I think it is worth building the habit of removing irrelevant ones, as well as updating or building on existing ones as appropriate.